File: 300. 最长上升子序列.cpp
/**
 * 注意这道题没有要求连续
 * dp[i]表示以nums[i]结尾的最长上升子序列长度；
 * 这道题求的一个数组的最长上升子序列长度，我们将问题分解，先求出以nums[0],num[1]...nums[n-1]结尾的最长升子序列长度，
 * 再求这n种情形中的最大值
 */

class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        
        int dp[10001];
        int maxLength=0;
        
        if(nums.size()==1)
            return 1;
        
        for(int i=0;i<nums.size();i++){
            dp[i]=1;
            for(int j=0;j<i;j++){
                if(nums[j]<nums[i]){
                    dp[i]=max(dp[i],dp[j]+1);
                }
                maxLength=max(maxLength,dp[i]);
            }
        }
        return maxLength;
    }
};

File: js/300. 最长上升子序列.js
/**
 * @param {number[]} nums
 * @return {number}
 */

/**
 * dp思路：dp[i]表示以nums[i]结尾的严格递增子序列最大长度
 * 思考dp[i],dp[j](j<i)之间的关系。
 */
var lengthOfLIS = function(nums) {
    let len=nums.length
    let dp=[]
    dp[0]=1
    let res=1

    for(let i=1;i<len;i++){
        // 初始化
        dp[i]=1
        for(let j=0;j<i;j++){
            if(nums[i]>nums[j])
                dp[i]=Math.max(dp[i],dp[j]+1)
        }
        res=Math.max(res,dp[i])
    }
    return res
};
